Util changes needed for VSD

thk123 opened the pull request:
Rebased version of #5205, with two additions:

 - tests for the const iterators added to base symbol table
 - stop definitions in symbol_table hiding the implemention in symbol_table_baset (risks these defintions going out of sync) 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Correct use of macros so that unit test work with USE_STD_STRING

This don't show up on travis because the unit tests are currently
not being built on Travis if USE_STD_STRING is set.  Originally by
Hannes Steffenhagen.

Commit:
Add an interface to iterate over symbol tables.

Patch originally by Chris Ryder.

Commit:
Add tests for iterators for symbol_tablet and symbol_table_baset

Check that both const and non-const iterators work as expected.

Commit:
Make the symbol_table_baset define the const iterators

Since the const iterators can use the base class directly, these are not
pure virtual - so derived classes can still provide an entirely
different mechanism for iterating over the symbols - though this is not
currently used.

hannes-steffenhagen-diffblue commented on unit/util/irep.cpp:
```diff
@@ -22,7 +22,7 @@ SCENARIO("irept_memory", "[core][utils][irept]")
 #endif
 
-#ifndef USE_STRING
+#ifndef USE_STD_STRING
```
Side note: I’d really love if we could just get rid of `USE_STD_STRING` already. I’m not entirely convinced we keep it around for anything other than nostalgia at this point.

martin-cs replied:
I believe the only use case for it is / was debugging problems in the `dtringt` implementation.  I am not sure how many issues it will find that have not already been found.

hannes-steffenhagen-diffblue commented on src/util/symbol_table_base.cpp:
```diff
@@ -60,6 +60,16 @@ void symbol_table_baset::show(std::ostream &out) const
 
+symbol_table_baset::const_iteratort symbol_table_baset::end() const
+{
+  return symbols.end();
```
96bd3ae8e86ac8712a662e80d710c05a64b8442c `s/interate/iterate` in the commit message.

Also this commit should be removed; Current develop already has begin/end for `symbol_tablet`.

hannes-steffenhagen-diffblue commented on unit/util/symbol_table.cpp:
```diff
@@ -17,14 +17,55 @@ TEST_CASE("Iterating through a symbol table", "[core][utils][symbol_tablet]")
 
-  int counter = 0;
-  for(auto &entry : symbol_table)
+  SECTION("Non const iterator")
```
This commit (a9afb9a25f3fd32e5e227bc9a91d7d5228bf38f6) can also be removed; symbol table iterators are tested in `unit/util/symbol_table.cpp`

thk123 replied:
I can see a single class to `begin()` but no call to end. This also tests the correct dispatching when you have a `symbol_tablet` but the compile time type is `symbol_table_baset`

hannes-steffenhagen-diffblue commented on src/util/symbol_table.h:
```diff
@@ -122,15 +122,8 @@ class symbol_tablet : public symbol_table_baset
-    return internal_symbols.end();
-  }
+  using symbol_table_baset::begin;
+  using symbol_table_baset::end;
```
develop `symbol_table_baset::begin/end` are both purely virtual, so can not be used in this way.

thk123 replied:
Discussed out of band: more consistent to have both const and non-const iterators be virtual and defined in the base class.

hannes-steffenhagen-diffblue reviewed (changes requested):
Other than the `USE_STD_STRING` change I don’t believe any of these changes should be merged as they already exist in some form or another on develop.

thk123 commented:
@hannes-steffenhagen-diffblue I've made the begin/end virtual in the base class as agreed, ready for [re-review](https://github.com/diffblue/cbmc/compare/acc46d626ecb385499286a86194eaa1527d1c09f..dca07dda7bb9a504cef9aca5d302d6f542f4267e)

martin-cs reviewed (approved):
Am happy once @hannes-steffenhagen-diffblue 's comments have been addressed.  I think that it will still work with the patch set.

hannes-steffenhagen-diffblue reviewed (approved):
I still think it’s a bit inconsistent we have a pure virtual mut `begin/end` but an implementation for const `begin/end`. But I’d not block on that if that’s the only complaint.

thk123 commented:
@peterschrammel if you get a min, would you be able to provide the code owner review for this short PR? You reviewed this originally in #5205, Martin addressed your comments (this is a rebased version)
